app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java

```diff
@@ -550,22 +556,103 @@ protected void deliverError(CommCareSetupActivity receiver,
     public boolean onCreateOptionsMenu(Menu menu) {
         super.onCreateOptionsMenu(menu);
         menu.add(0, MODE_ARCHIVE, 0, Localization.get("menu.archive")).setIcon(android.R.drawable.ic_menu_upload);
+        menu.add(0, MODE_SMS, 1, Localization.get("menu.sms")).setIcon(android.R.drawable.ic_menu_upload);
         return true;
     }
 
     @Override
     public boolean onPrepareOptionsMenu(Menu menu) {
         super.onPrepareOptionsMenu(menu);
-        
+
         return true;
     }
-    
+
+    /**
+     * Scan SMS messages for texts with profile references.
+     * @param installAutomatically install automatically if reference is found
+     */
+    public void performSMSInstall(boolean installAutomatically){
+        try {
+            String profileLink = this.scanSMSLinks();
+            if (profileLink != null) {
+                // we found a valid profile link, either start install automatically
+                // or move to READY_TO_INSTALL state
+                Log.v("install", "Performing SMS install with link : " + incomingRef);
+                incomingRef = profileLink;
+                if (installAutomatically) {
+                    startResourceInstall();
+                } else {
+                    uiState = UiState.READY_TO_INSTALL;
+                    Toast.makeText(this, Localization.get("menu.sms.ready"), Toast.LENGTH_LONG).show();
+                }
+            } else {
+                // only notify if this was manually triggered, since most people won't use this
+                if(installAutomatically) {

```

Comment:
This instance of `else { if (...) { ... ` can be collapsed to `else if (...) { ...`